### DISCUSSION TITLE
Clearing credentials in eeprom with arduino_homekit_server.h

### DISCUSSION THREAD

**Discussion by FRANAIRBUS** [Hieromon/AutoConnect#506] [How can I?] [answered]
Hi, great job on the autoconnet library I use it in many of my projects

Now I encounter a problem of deleting credentials when using the arduino_homekit_server.h library together and the wifi connection credentials are deleted and at the same time the pairing of homekit when restarting, if the equipment does not restart it works correctly but when restarting the credentials are deleted and pairing with homekit. I think it is a problem in the use of the eeprom.

someone is so kind to tell me how to change the use of the eeprom or the solution to this problem?

as per autoconnect manual Default offset value is 0 for credentials On ESP8266 or ESP32 

And in homekit library

// ESP82666  spi_flash_geometry.h
// FLASH_SECTOR_SIZE 0x1000 = 4096B
// EEPROM use _sector = _EEPROM_start
// the "_EEPROM_start" value is provided in tools/sdk/ld/eagle.flash.**.ld

// We use the EEPROM address for this storage.c in Arduino environment
// EEPROM = one SPI_FLASH_SEC_SIZE = 0x1000 = 4096B
// SPI_FLASH_SEC_SIZE = 4096B
// sizeof(pairing_data_t) = 80B
// MAX_PAIRINGS = 16
// PAIRINGS_OFFSET = 128(address)
// so max use address of (128 + 80 * 16) = 1408 B

// This storage use [0, 1408) of EEPROM, leave [1408, 4096) for user to use safely.
// Leave the FS(file system) for user to use freely.

I'm a bit lost on this topic
the example used is
[mqttRSSI](https://github.com/Hieromon/AutoConnect/tree/master/examples/mqttRSSI)
With
https://github.com/Mixiaoxiao/Arduino-HomeKit-ESP8266

Thank you

**Comment by Hieromon** [comment, marked as answer]
@FRANAIRBUS Thank you for using AutoConnect.

According to the Arduino-HomeKit-ESP8266 guidance, their library also uses EEPROM. It declares that it natively uses the SDK API (spi_flash_read, spi_flash_write) without going through the ESP8266 EEPROM library of Arduino ESP8266 core. On the other hand, AutoConnect accesses the EEPROM via the ESP8266 Arduino core library; since the ESP8266 EEPROM library is a wrapper for spi_flash_read/spi_flash_write, so there will be no conflict between AutoConnect and Arduino-HomeKit-ESP8266 due to different access methods.

The cause of the loss of AutoConnect's stored credentials is probably an EEPROM address conflict. Here is a quote from [README.md of Arduino-HomeKit-ESP8266](https://github.com/Mixiaoxiao/Arduino-HomeKit-ESP8266#storage).

> The EEPROM is 4096B in ESP8266, this project uses max [0, 1408B).

This address overlaps with the **default address** of EEPROM storage for AutoConnect credentials. To avoid this overlap, you can use the [AutoConnectConfig::boundaryOffset](https://hieromon.github.io/AutoConnect/apiconfig.html#boundaryoffset) setting. 
You can refer to the [documentation](https://hieromon.github.io/AutoConnect/adcredential.html#move-the-saving-area-of-eeprom-for-the-credentials) for more information, but the implementation code would probably look something like follows:

```cpp
#include <Arduino.h>
#include <ESP8266WiFi.h>
#include <ESP8266WebServer.h>
#include <AutoConnect.h>
#include <arduino_homekit_server.h>

AutoConnect portal;
AutoConnectConfig acConfig;

extern "C" homekit_server_config config;

void setup() {
  delay(1000);
  acConfig.boundaryOffset = 2048; // The actual capacity required by Arduino-HomeKit is 1408 bytes, but we expect a little leeway.
  portal.config(acConfig);
  portal.begin();
  arduino_homekit_setup(&config);
}

void loop() {
  arduino_homekit_loop();
  // delay(10); <- It is wise to exclude this line, which is exemplified in Arduino-HomeKit, as it may interfere with TCP session interrupts.

  portal.handleClient();
}
```

**NOTES**
In the AutoConnect release test, AutoConnectConfig::boundaryOffet has not tested with a large value such as 2048. Therefore, it is possible that this setting does not work correctly. If you encounter problems using the above workaround, please post again with the conditions under which it occurred. I will look into ways to address it.

**Reply by FRANAIRBUS** [reply]
Hello @Hieromon thank you very much for your prompt and quick response,
the solution acConfig.boundaryOffset = 2048; is correct :) as always your code manages to solve many problems.

once connected with AutoConnect and paired with homekit everything seems fine and the credentials are not deleted, but there are 2 problems with the homekit library that I still have to investigate

1º in the first start of homekit it deletes so I have to re-record the wifi credentials.

10:40:57.729 -> >>> [ 121581] HomeKit: Starting server
10:40:57.729 -> >>> [ 121585] HomeKit: Formatting HomeKit storage at 0xffb000
10:40:57.776 -> >>> [ 121624] HomeKit: Generated new accessory ID: C5:1C:AF:11:FF:5B
10:40:57.809 -> >>> [ 121683] HomeKit: Generated new accessory key
10:40:57.853 -> >>> [ 121688] HomeKit: Preiniting pairing context
10:40:57.853 -> >>> [ 121693] HomeKit: Using user-specified password: 111-11-111
10:40:57.853 -> >>> [ 121713] HomeKit: Call s_mp_exptmod in integer.c, original winsize 6
10:41:04.548 -> >>> [ 128381] HomeKit: Call s_mp_exptmod in integer.c, original winsize 5
10:41:07.960 -> >>> [ 131797] HomeKit: Preinit pairing context success
10:41:07.960 -> >>> [ 131802] HomeKit: Configuring MDNS
10:41:07.960 -> >>> [ 131806] HomeKit: MDNS.begin: LEYNA4, IP: 192.168.20.252

2º once I put the credentials for the second time and paired to homekit it is not deleted and everything seems to work even restarting esp8266 but if the wifi router goes down and AutoConnect searches and reconnects with the saved credentials, homekit does not work without rebooting esp8266 .

abstract:
acConfig.boundaryOffset = 2048; it's ok good job
arduino_homekit_server library should be investigated further to understand its operation.
I can't find
delay(10); <- It is wise to exclude this line, which is exemplified in Arduino-HomeKit, as it may interfere with TCP session interrupts.

In my code:

    void loop() {
    if (WiFi.status() != WL_CONNECTED) {
    if (millis() - tm > SCAN_INTERVAL) {
        
        Serial.println("SEARCHING NETWORK");
        portal.begin();
        yield();
        tm = millis();
     }
     }
    portal.handleClient();
    my_homekit_loop();
    }
    
    }




Great job and excuse my english

**Reply by Hieromon** [reply]
@FRANAIRBUS,

> 1º in the first start of homekit it deletes so I have to re-record the wifi credentials.
> 2º once I put the credentials for the second time and paired to homekit it is not deleted and everything seems to work even restarting esp8266 but if the wifi router goes down and AutoConnect searches and reconnects with the saved credentials, homekit does not work without rebooting esp8266 .

I found your mention above odd, so I have traced the Arduino-HomeKit code lightly. Indeed it appears to be block erasing the EEPROM area after the instantiation of the WiFiServer within the initialization sequence.
https://github.com/Mixiaoxiao/Arduino-HomeKit-ESP8266/blob/c862e2e5f113900765522fa9a0091b00fe27975b/src/arduino_homekit_server.cpp#L3347-L3354

In conclusion, you may be able to work around the problem by saving AutoConnect credentials to the filesystem before initializing the Arduino-HomeKit as `homekit_server_init` call.
The evaluation code below shows only the saving of credentials to the file system. You can incorporate this snippet into your Arduino-HomeKit sketch to verify its behavior.

```cpp
#include <Arduino.h>
#include <ESP8266WiFi.h>
#include <ESP8266WebServer.h>
#include <AutoConnect.h>
#include <LittleFS.h>
#include <arduino_homekit_server.h>

AutoConnect portal;
AutoConnectConfig acConfig;

extern "C" homekit_server_config config;
const uint16_t  CREDENTIAL_OFFSET = 2048;
const char* CREDENTIAL_DIR = "/";
const char* CREDENTIAL_FILEPREFIX = "ac_";

bool restoreCredentials(const uint16_t offset, const char* dir, const char* prefix) {
  if (!LittleFS.begin()) {
    Serial.println("Fails LittleFS start");
    return false;
  }

  Dir root = LittleFS.openDir(dir);
  while (root.next()) {
    if (root.fileName().startsWith(prefix)) {
      File  file = root.openFile("r");
      if (file) {
        AutoConnectCredential credt(offset);
        station_config_t  sta_config;
        if (file.readBytes((char*)&sta_config, sizeof(sta_config))) {
          if (credt.save(&sta_config))
            Serial.printf("Credential:%s restored\n", (char*)sta_config.ssid);
          else
            Serial.printf("Failed to restore credential %s\n", (char*)sta_config.ssid);
        }
        else
          Serial.printf("Failed to load credential %s\n", file.name());
        file.close();
      }
    }
  }

  LittleFS.end();
  return true;
}

bool saveCredentials(const uint16_t offset, const char* dir, const char* prefix) {
  if (!LittleFS.begin()) {
    Serial.println("Fails LittleFS start");
    return false;
  }

  AutoConnectCredential credt(offset);
  station_config_t  sta_config;

  bool  rc = true;
  uint8_t entries = credt.entries();
  Serial.printf("%" PRIu8 " available credential(s)\n", entries);
  for (uint8_t i = 0; i < entries; i++) {
    credt.load(i, &sta_config);
    Serial.printf("Credential(%" PRIu8 ") %s, ", i, (char*)sta_config.ssid);
    String  cfName = String(dir) + String(prefix) + String((char*)sta_config.ssid);
    File  credential = LittleFS.open(cfName, "w");
    if (credential.write((uint8_t*)&sta_config, sizeof(station_config_t)))
      Serial.printf("saved to %s\n", cfName.c_str());
    else {
      Serial.println("failed to save");
      rc = false;
    }
    credential.close();
  }

  return rc;
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();

  restoreCredentials(CREDENTIAL_OFFSET, CREDENTIAL_DIR, CREDENTIAL_FILEPREFIX);

  acConfig.autoReconnect = true;  // Allow background reconnect within AutoConnect::handleClient.
  acConfig.reconnectInterval = 1;  // A reconnection interval of 30 seconds applies.
  acConfig.retainPortal = true;
  acConfig.boundaryOffset = CREDENTIAL_OFFSET;
  portal.config(acConfig);

  if (portal.begin())
    saveCredentials(CREDENTIAL_OFFSET, CREDENTIAL_DIR, CREDENTIAL_FILEPREFIX);

  arduino_homekit_setup(&config);
}

void loop() {
  // AutoConnectConfig::autoReconnect's utility makes the following code block unnecessary.
  // if (WiFi.status() != WL_CONNECTED) {
  //   if (millis() - tm > SCAN_INTERVAL) {
  //     Serial.println("SEARCHING NETWORK");
  //     portal.begin();
  //     yield();
  //     tm = millis();
  //   }
  // }

  portal.handleClient();
  arduino_homekit_loop();
}
```

The above snippet saves AutoConnect credentials to **LittleFS**. One file is created per credential, with the SSID as the file name. However, there are a few caveats to applying the above snippet to your sketch.

1. SPIFFS is already deprecated in the ESP8266arduino core; Arduino-HomeKit is aware of the existence of the SPIFFS file system in its initialization sequence, but it may not be compatible with LittleFS.
2. It is not possible to save credentials that are hidden SSIDs. In this case, you need to change the file name from SSID to BSSID, etc.

### In addition

Your request inspired me to imagine your implementation. And I noticed that AutoConnect lacks the ability to export credentials in bulk. I decided to create one file for each SSID to deal with this issue, but this strategy is a bit redundant. Bulk transfer is inherently preferable.
I will support bulk export and restore of credentials to the file system in the next update turn of AutoConnect enhancement. With its extended implementation, I expect the sketch to look like this:

```cpp
AutoConnect portal;
AutoConnectConfig config;

void saveCredentials(IPAddress& ip) {
  portal.export("AC_CREDENTIAL");  // save whole credentials to AC_CREDENTIAL file
}

void setup() {
  portal.onConnect(saveCredentials);  // An onConnect exit is already supported

  config.boundaryOffset = 2048;
  portal.config(config);
  portal.import("AC_CREDENTIAL");  // restore whole credentials from AC_CREDENTIAL file
  portal.begin();
}

void loop() {
  portal.handleClient();
}
```

`AutoConnect::export` and `AutoConnect::import` perform batch saving and recovery of credentials to the file system, respectively. There will be only one file used for saving, which may be specified optionally. Also, I would like it to be able to specify other than the default file system instance.
I will now start working on the details of the specifications, so if you have any requests about this, please let me know.

**Reply by FRANAIRBUS** [reply]
wow awesome
you are a genius and I am glad that you think of this implementation.

my full build as a smart thermostat includes

#include pushbuttons.h          // function pushbuttons for relay, lights, thermostat and others
#include pin_in_out.h             //configure input and output pins
#include msql.h                      // send all data and status to a database
#include myblynk.h                // "deprecated" blynk app will end up switching to my own app.
#include mythermostat.h       // smart thermostat including probe calibration, setpoint, etc
#include probes.h                  // dht, DS18B20, MQ-7 and others
#include myfunctions.h

this build has been working fine with your library for a long time with the example of mqttRSSI_FS with 4 pages on autoconnect aux and ota.

now i want to update it without the need to load separate json files and would like to include this homekit option as well.

thank you very much again.
I'll try your implementation.


// AutoConnectConfig::autoReconnect's utility makes the following code block unnecessary.
  // if (WiFi.status() != WL_CONNECTED) {
  // if (millis() - tm > SCAN_INTERVAL) {
  // Serial.println("SEARCHING NETWORK");
  // portal.begin();
  // yield();
  // tm = millis();
  // }
  // }

I think I remember that I implemented it so that the thermostat loop continues to work without wifi

I will also try if it is correct

I'll ask you in another discussion. what do you think should be the correct implementation for multiple auto-connected configuration pages?

autoconnect elements or json to update ota with a single precompiled .bin

**Reply by FRANAIRBUS** [reply]
Good morning,
the solution of saving AutoConnect credentials in the file system is correct and stable.
I have tested and everything works fine.

just a small detail.
extern "C" homekit_server_config config; // it does not work
extern "C" homekit_server_config_t config; // it works

On the other hand, as I commented, if the router off homekit does not reconnect and does not have automatic reconnection.

solution found:

https://github.com/Mixiaoxiao/Arduino-HomeKit-ESP8266/issues/139#issuecomment-940078639

comment
https://github.com/Mixiaoxiao/Arduino-HomeKit-ESP8266/blob/c862e2e5f113900765522fa9a0091b00fe27975b/src/arduino_homekit_server.cpp#L3165

// MDNS.close();
Line 3165 within
src/arduino_homekit_server.cpp

thanks and great job

an idea? I like it
AutoConnectConfig acConfig;
for your examples instead of config since the latter is used by many other libraries and for someone new it generates an error
